Code/Framework/AzCore/AzCore/Task/TaskGraph.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/Task/TaskGraph.h>
+
+#include <AzCore/Task/TaskExecutor.h>
+
+namespace AZ
+{
+    using Internal::CompiledTaskGraph;
+
+    void TaskToken::PrecedesInternal(TaskToken& comesAfter)
+    {
+        AZ_Assert(!m_parent.m_submitted, "Cannot mutate a TaskGraph that was previously submitted.");
+
+        // Increment inbound/outbound edge counts
+        m_parent.m_tasks[m_index].Link(m_parent.m_tasks[comesAfter.m_index]);
+
+        m_parent.m_links[m_index].emplace_back(comesAfter.m_index);
+
+        ++m_parent.m_linkCount;
+    }
+
+    TaskGraph::~TaskGraph()
+    {
+        if (m_retained && m_compiledTaskGraph)
+        {
+            // This job graph has already finished and we are potentially responsible for its destruction
+            if (m_compiledTaskGraph->Release() == 0)
+            {
+                azdestroy(m_compiledTaskGraph);
+            }
+        }
+    }
+
+    void TaskGraph::Reset()
+    {
+        AZ_Assert(!m_submitted, "Cannot reset a job graph while it is in flight");
+        if (m_compiledTaskGraph)
+        {
+            azdestroy(m_compiledTaskGraph);
+            m_compiledTaskGraph = nullptr;
+        }
+        m_tasks.clear();
+        m_links.clear();
+        m_linkCount = 0;
+    }
+
+    void TaskGraph::Submit(TaskGraphEvent* waitEvent)
+    {
+        SubmitOnExecutor(TaskExecutor::Instance(), waitEvent);
+    }
+
+    void TaskGraph::SubmitOnExecutor(TaskExecutor& executor, TaskGraphEvent* waitEvent)
+    {
+        if (!m_compiledTaskGraph)
+        {
+            m_compiledTaskGraph = aznew CompiledTaskGraph(AZStd::move(m_tasks), m_links, m_linkCount, m_retained ? this : nullptr);
+        }
```

Comment:
Aren't we ending up with a runtime alloc when we reset/resubmit the task graph as the compiled task graph gets destroyed in the `Reset()`?

Reply:
Would it make sense to introduce a dirty flag to the task graph and enable that in case a task/dependency got added/removed and only recreate the compiled task in such case?

Reply:
You don't actually call `Reset` unless you actually want to rebuild the graph again from the beginning. The graph is automatically retained and you can just resubmit it each frame (changing whatever data is in the refs/pointers captured by the lambda itself).

We could do dirty tracking I suppose, I guess I didn't immediately see a use case for it and didn't want a larger API than necessary to start. Did you have a specific use case in mind, or you would mind deferring that to later?